packages/react-components/react-utilities/src/hooks/useBrowserTimer.ts

```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+
+type UseBrowserTimerSetter =
+  | ((fn: () => void, duration?: number, ...args: Record<string, unknown>[]) => number)
+  | ((fn: () => void) => number);
+type UseBrowserTimerCancel = ((timerId: number) => void) | (() => void);
+
+/**
+ * @internal
+ * Helper to manage a browser timer.
+ * Ensures that the timer isn't set multiple times at once,
+ * and is cleaned up when the component is unloaded.
+ *
+ * @param setTimer - The timer setter function
+ * @param cancelTimer - The timer cancel function
+ * @returns A pair of [setTimer, cancelTimer] that are stable between renders.
+ *
+ * @example
+ * const [setTimer, cancelTimer] = useBrowserTimer(setTimeout, cancelTimeout);
+ *
+ * setTimer(() => console.log('Hello world!'), 1000);
+ * cancelTimer();
+ */
+export function useBrowserTimer<TSetter extends UseBrowserTimerSetter, TCancel extends UseBrowserTimerCancel>(
+  setTimer: TSetter,
+  cancelTimer: TCancel,
+) {
+  const [timeout] = React.useState(() => ({
+    id: undefined as number | undefined,
+    set: (fn: () => void, delay?: number) => {
+      timeout.cancel();
+      timeout.id = delay ? setTimer(fn, delay) : setTimer(fn);
+      return timeout.id;
+    },
+    cancel: () => {
+      if (timeout.id !== undefined) {
+        cancelTimer(timeout.id);
+        timeout.id = undefined;
+      }
+    },
+  }));
+
```

Comment:
nit:

```suggestion
export function useBrowserTimer(
  setTimer: (fn: () => void, delay?: number) => number,
  cancelTimer: (id: number) => void,
) {
  const id = React.useRef<number | undefined>(undefined);
  const set = React.useCallback(
    (fn: () => void, delay?: number) => {
      id.current = setTimer(fn, delay);
      return id.current;
    },
    [setTimer],
  );
  const cancel = React.useCallback(() => {
    if (id.current !== undefined) {
      cancelTimer(id.current);
      id.current = undefined;
    }
  }, [cancelTimer]);
  // Clean up the timeout when the component is unloaded
  React.useEffect(() => cancel, [cancel]);

  return [set, cancel] as const;
}
```

You can refactor this to use just a ref instead of a state, and also no need for generics, this signature is enough to support both `raf` and `setTimeout` callbacks

Reply:
That makes sense. That might be a little faster than the previous iteration of useTimeout. I made only a little change to the typings, as setTimeout and setInterval accept a spread of parameters, so with this extra type we cover that.